esp/services/ws_codesign/ws_codesignService.cpp

```diff
@@ -0,0 +1,100 @@
+/*##############################################################################
+
+    HPCC SYSTEMS software Copyright (C) 2019 HPCC Systems®.
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+############################################################################## */
+
+#include "ws_codesignService.hpp"
+#include "jutil.hpp"
+
+Cws_codesignEx::Cws_codesignEx()
+{
+}
+
+Cws_codesignEx::~Cws_codesignEx()
+{
+}
+
+void Cws_codesignEx::init(IPropertyTree *cfg, const char *process, const char *service)
+{
+    if(cfg == nullptr)
+        throw MakeStringException(-1, "Cannot initialize Cws_codesignEx, cfg is NULL");
+
+    StringBuffer xpath;
+    xpath.appendf("Software/EspProcess[@name=\"%s\"]/EspService[@name=\"%s\"]", process, service);
+    m_serviceCfg.setown(cfg->getPropTree(xpath.str()));
+}
+
+bool Cws_codesignEx::onSign(IEspContext &context, IEspSignRequest &req, IEspSignResponse &resp)
+{
+    resp.setRetCode(-1);
+
+    const char* keyid = req.getKeyID();
+    const char* text = req.getText();
+    if (!keyid || !*keyid || !text || !*text)
+    {
+        resp.setErrMsg("Please provide both KeyID and Text");
+        return false;
+    }
+    int keylen = strlen(keyid);
+    if(keylen != 16 && keylen != 40)
+    {
+        resp.setErrMsg("Invalid KeyID");
+        return false;
+    }
+
+    StringBuffer cmd, output, errmsg;
+
+    cmd.appendf("gpg --list-secret-keys --with-keygrip %s", keyid);
+    int ret = runExternalCommand(output, errmsg, cmd.str(), "");
+    if (ret != 0 || strstr(output.str(), keyid) == nullptr)
+    {
+        resp.setErrMsg("Key not found");
+        return false;
+    }
+
+    StringBuffer keygrip;
+    if (auto kgptr = strstr(output.str(), "Keygrip = "))
```

Comment:
style: I think this would be clearer, if the assignment was on a separate line from an if, and less similar to an error prone assignment within an if condition.